src/fs/metadata_log.hh

```diff
@@ -0,0 +1,252 @@
+
+    future<> flush_curr_cluster_and_change_it_to_new_one();
+
+    template<class... Args>
+    future<> append_ondisk_entry(Args&&... args) {
+        return repeat([this, args...] {
+            auto append_res = _curr_cluster_buff->append(args...);
+            switch (append_res) {
+            case metadata_to_disk_buffer::APPENDED:
+                return make_ready_future<stop_iteration>(stop_iteration::yes);
+
+            case metadata_to_disk_buffer::TOO_BIG: {
+                return flush_curr_cluster_and_change_it_to_new_one().then([] {
+                    return stop_iteration::no;
+                });
+            }
+            }
+
+            __builtin_unreachable();
+        });
+    }
+
+    enum class path_lookup_error {
+        NOT_ABSOLUTE, // a path is not absolute
+        NO_ENTRY, // no such file or directory
+        NOT_DIR, // a component used as a directory in path is not, in fact, a directory
+    };
+
+    std::variant<inode_t, path_lookup_error> path_lookup(const sstring& path) const;
+
+    future<inode_t> futurized_path_lookup(const sstring& path) const;
+
+public:
+    template<class Func>
+    future<> iterate_directory(const sstring& dir_path, Func func) {
+        static_assert(std::is_invocable_r_v<future<>, Func, const sstring&> or std::is_invocable_r_v<future<stop_iteration>, Func, const sstring&>);
+        auto convert_func = [&]() -> decltype(auto) {
+            if constexpr (std::is_invocable_r_v<future<stop_iteration>, Func, const sstring&>) {
+                return std::move(func);
+            } else {
+                return [func = std::move(func)]() -> future<stop_iteration> {
+                    return func().then([] {
+                        return stop_iteration::no;
+                    });
+                };
+            }
+        };
+        return futurized_path_lookup(dir_path).then([this, func = convert_func()](inode_t dir_inode) {
+            return do_with(std::move(func), sstring {}, [this, dir_inode](auto& func, auto& prev_entry) {
+                auto it = _inodes.find(dir_inode);
+                if (it == _inodes.end()) {
+                    return now(); // Directory disappeared
+                }
+                if (not std::holds_alternative<inode_info::directory>(it->second.contents)) {
+                    return make_exception_future(path_component_not_directory());
+                }
+
+                return repeat([this, dir_inode, &prev_entry, &func] {
+                    auto it = _inodes.find(dir_inode);
+                    if (it == _inodes.end()) {
+                        return make_ready_future<stop_iteration>(stop_iteration::yes); // Directory disappeared
+                    }
+                    if (not std::holds_alternative<inode_info::directory>(it->second.contents)) {
+                        return make_ready_future<stop_iteration>(stop_iteration::yes); // Directory became a file (should not happen, but safe-check)
```

Comment:
I think it is not probable, but possible so I would remove that part "should not happen..." because it could be misleading and suggest that it is impossible unless there is a bug in code.

Reply:
I removed "(should not happen, but safe-check)" part.